onnxruntime/test/framework/inference_session_test.cc

```diff
@@ -498,6 +499,59 @@ TEST(InferenceSessionTests, TestModelSerialization) {
   ASSERT_TRUE(session_object_emptyValidation.Initialize().IsOK());
 }
 
+TEST(InferenceSessionTests, TestLoadCancellation) {
+  SessionOptions so;
+  so.session_logid = "InferenceSessionTests.TestLoadCancellation";
+
+  {
+    // Explicit cancel during load, small model is fine
+    const PathString model_uri = ORT_TSTR("testdata/constant_floats.onnx");
+    InferenceSession session_object{so, GetEnvironment()};
+    *so.load_cancellation_flag = true;
+    ASSERT_FALSE(session_object.Load(model_uri).IsOK());
+  }
+  {
+    // Explicit cancel during initialize, small model is fine
+    const PathString model_uri = ORT_TSTR("testdata/constant_floats.onnx");
+    *so.load_cancellation_flag = false;
+    InferenceSession session_object{so, GetEnvironment()};
+    ASSERT_STATUS_OK(session_object.Load(model_uri));
+    *so.load_cancellation_flag = true;
+    ASSERT_FALSE(session_object.Initialize().IsOK());
+  }
+
+  {
+    // Large model that takes time to load
+    const PathString test_model = ORT_TSTR("testdata/transformers/tiny_gpt2_beamsearch.onnx");
+    auto terminator = [&so]() {
+      std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+      *so.load_cancellation_flag = true;
+      return;
+    };
+
+    std::packaged_task<void()> task{terminator};
+    std::future<void> terminator_result = task.get_future();
+    std::thread terminator_thread{std::move(task)};
+
+    bool terminated = false;
+    try {
+      InferenceSession session_object{
+          so, GetEnvironment(), test_model};
+
+    } catch (const std::exception& ex) {
+      std::string m{ex.what()};
+      terminated = m.find("user request") != std::string::npos;
+    }
+    // done with the thread
+    terminator_thread.join();
+
+    // call get to propagate any exception
+    terminator_result.get();
+
+    ASSERT_TRUE(terminated);
```

Comment:
this test relies on a data race on reading and writing the flag. can we test it in a way that does not rely on a certain interleaving of unsynchronized threads?

Reply:
I think this comment still applies. it is possible (perhaps unlikely with the current hardware) for loading to finish successfully before the terminator_thread gets around to setting the load cancellation flag.